File: ruby/prices.rb
#!/usr/bin/env ruby

require 'csv'
require 'date'

class Product
  def self.create(supplier_id, product_code, description, delivery_date, cost_price, unit_count)
    if product_code >= 1000 and product_code <= 1999
      Fruit.create(supplier_id, product_code, description, delivery_date, cost_price, unit_count)
    else
      raise "Bad product code: #{product_code} #{description}"
    end
  end
  
  def initialize(supplier_id, product_code, description, delivery_date, cost_price, unit_count)
    @supplier_id   = supplier_id
    @product_code  = product_code
    @description   = description
    @delivery_date = delivery_date
    @cost_price    = cost_price
    @unit_count    = unit_count
  end

  def shelf_days
    7
  end

  def sell_by_date
    if @supplier_id == 32
      days_to_add = shelf_days - 3
    else
      days_to_add = shelf_days
    end
    @delivery_date + days_to_add
  end

  def markup_percentage
    50
  end

  def markup
    @cost_price*(markup_percentage/100.0)
  end

  def sell_price_in_rands
    (@cost_price + markup)/100.0
  end

  def label_sell_price
    'R' + "% 8.2f" % sell_price_in_rands
  end

  def label_sell_by_date
    sell_by_date.strftime('%Y/%m/%d')
  end

  def label_description
    @description[0, 31]
  end
  
  def write_pricefile(file)
    @unit_count.times {file.puts label_sell_price + label_sell_by_date + label_description}
  end
end

class Fruit < Product
  def self.create(supplier_id, product_code, description, delivery_date, cost_price, unit_count)
    if product_code >= 1100 and product_code <= 1199
      Apple.new(supplier_id, product_code, description, delivery_date, cost_price, unit_count)
    elsif product_code >= 1200 and product_code <= 1299
      Banana.new(supplier_id, product_code, description, delivery_date, cost_price, unit_count)
    elsif product_code >= 1300 and product_code <= 1399
      Berry.new(supplier_id, product_code, description, delivery_date, cost_price, unit_count)
    else
      raise "Bad Fruit product code: #{product_code} #{description}" 
    end
  end
end

class Apple < Fruit
  def markup_percentage
    40
  end
  def shelf_days
    14
  end
end

class Banana < Fruit
  def markup_percentage
    35
  end
  def shelf_days
    5
  end
end

class Berry < Fruit
  def markup_percentage
    55
  end
end


def to_price_file(line, file)
  CSV.parse(line) do |row|
    supplier_id   = Integer(row[0])
    product_code  = Integer(row[1])
    description   = row[2]
    delivery_date = Date.parse(row[3])
    cost_price    = Integer(row[4])
    unit_count    = Integer(row[5])
    product = Product.create(supplier_id, product_code, description, delivery_date, cost_price, unit_count)
    product.write_pricefile(file)
  end
end

def main()
  inputfile  = File.new('../produce.csv', 'r')
  outputfile = File.new('pricefile.txt', 'w')

  inputfile.each_line("\n") do |line|
    if inputfile.lineno > 1
      to_price_file(line, outputfile)
    end
  end
end

main()

File: erlang/prices.erl
-module(prices).

-export([start/0]).

-record(product, {product_type,
                  product_code,
                  description,
                  delivery_date,
                  cost_price,
                  unit_count,
                  sell_price,
                  sell_by_date}).

start() ->
  {ok, InputFile}  = file:open("../produce.csv", [read]),
  {ok, OutputFile} = file:open("pricefile.txt", [write]),
  {ok, _Header} = file:read_line(InputFile),
  ok = process_file(InputFile, OutputFile),
  ok.

process_file(InputFile, OutputFile) ->
  case file:read_line(InputFile) of
    {ok, Line} ->
      process_file(InputFile, OutputFile, Line);
    eof ->
      ok
  end.

process_file(InputFile, OutputFile, Line) ->
  to_price_file(string:strip(Line, right, $\n), OutputFile),
  process_file(InputFile, OutputFile).

to_price_file(Line, OutputFile) ->
  Fields = csv:parse_line(Line),
  [_SupplierID, ProductCode, Description, DeliveryDate, CostPrice, UnitCount] = Fields,
  Product = calculate_product(list_to_integer(ProductCode), Description, parse_date(DeliveryDate),
                              list_to_integer(CostPrice), list_to_integer(UnitCount)),
  ok = write_pricefile(Product, OutputFile),
  ok.

calculate_product(ProductCode, Description, DeliveryDate, CostPrice, UnitCount) ->
  Product    = simple_product(ProductCode, Description, DeliveryDate, CostPrice, UnitCount),
  SellPrice  = sell_price(Product#product.product_type, CostPrice),
  SellByDate = sell_by_date(Product#product.product_type, DeliveryDate),
  Product#product{sell_price = SellPrice, sell_by_date=SellByDate}.

sell_price(ProductType, CostPrice) ->
  CostPrice + markup(ProductType, CostPrice).

sell_by_date(ProductType, DeliveryDate) ->
  date_add_days(DeliveryDate, shelf_days(ProductType)).

simple_product(ProductCode, Description, DeliveryDate, CostPrice, UnitCount) ->
  #product{product_type  = product_type(ProductCode),
           product_code  = ProductCode,
           description   = Description,
           delivery_date = DeliveryDate,
           cost_price    = CostPrice,
           unit_count    = UnitCount}.

markup(apple, CostPrice) ->
  CostPrice * (40/100.0);
markup(banana, CostPrice) ->
  CostPrice * (35/100.0);
markup(berry, CostPrice) ->
  CostPrice * (55/100.0);
markup(_ProductType, CostPrice) ->
  CostPrice * (50/100.0).

shelf_days(apple)        -> 14;
shelf_days(banana)       -> 5;
shelf_days(_ProductType) -> 7.

product_type(ProductCode) when ((ProductCode >= 1100) and (ProductCode =< 1199)) ->
  apple;
product_type(ProductCode) when ((ProductCode >= 1200) and (ProductCode =< 1299)) ->
  banana;
product_type(ProductCode) when ((ProductCode >= 1300) and (ProductCode =< 1399)) ->
  berry.

date_add_days(Date, Days) ->
  GregDays = calendar:date_to_gregorian_days(Date),
  calendar:gregorian_days_to_date(GregDays+Days).

parse_date(DateString) ->
  Tokens = string:tokens(DateString, "/"),
  [Year, Month, Day] = [list_to_integer(Str) || Str <- Tokens],
  {Year, Month, Day}.

date_to_string(Date) ->
  {Y, M, D} = Date,
  Year  = string:right(integer_to_list(Y), 4, $0),
  Month = string:right(integer_to_list(M), 2, $0),
  Day   = string:right(integer_to_list(D), 2, $0),
  io_lib:format("~s/~s/~s", [Year,Month,Day]).

write_pricefile(Product, OutputFile) ->
  SellPriceRands   = io_lib:format("~.2f", [Product#product.sell_price/100]),
  [SpaceChar] = " ",
  LabelSellPrice   = "R"++ string:right(SellPriceRands, 4, SpaceChar),
  LabelSellByDate  = date_to_string(Product#product.sell_by_date),
  LabelDescription = string:substr(Product#product.description, 1, 31),
  Line = LabelSellPrice ++ LabelSellByDate ++ LabelDescription ++ "\n",
  Lines = [Line || _Count <- lists:seq(1,Product#product.unit_count)],
  ok = file:write(OutputFile, Lines),
  ok.
